cardano-testnet/src/Testnet/Types.hs

```diff
@@ -101,22 +114,35 @@ data TestnetRuntime = TestnetRuntime
   { configurationFile :: !(NodeConfigFile In)
   , shelleyGenesisFile :: !FilePath
   , testnetMagic :: !Int
-  , poolNodes :: ![PoolNode]
+  , testnetNodes :: ![TestnetNode]
   , wallets :: ![PaymentKeyInfo]
   , delegators :: ![Delegator]
   }
 
-poolSprockets :: TestnetRuntime -> [Sprocket]
-poolSprockets = fmap (nodeSprocket . poolRuntime) . poolNodes
+testnetSprockets :: TestnetRuntime -> [Sprocket]
+testnetSprockets = fmap (nodeSprocket . testnetNodeRuntime) . testnetNodes
+
+allNodes :: TestnetRuntime -> [NodeRuntime]
+allNodes = fmap testnetNodeRuntime . testnetNodes
+
+spoNodes :: TestnetRuntime -> [NodeRuntime]
+spoNodes = fmap testnetNodeRuntime . filter isTestnetNodeSpo . testnetNodes
+
+relayNodes :: TestnetRuntime -> [NodeRuntime]
+relayNodes = fmap testnetNodeRuntime . filter (not . isTestnetNodeSpo) . testnetNodes
 
-data PoolNode = PoolNode
-  { poolRuntime :: NodeRuntime
-  , poolKeys :: PoolNodeKeys
+data TestnetNode = TestnetNode
+  { testnetNodeRuntime :: !NodeRuntime
+  , poolKeys :: Maybe SpoNodeKeys -- ^ Keys are only present for SPO nodes
```

Comment:
We should move `poolKeys` into `NodeRuntime` if possible.

Reply:
I wasn't sure about that. `NodeRuntime` is more about the system process itself like sockets, system handles etc. `TestnetNode` is rather for describing logical layer, like keys here. 

With the amount of information we have now it can be one datatype I guess.

Reply:
https://github.com/IntersectMBO/cardano-node/pull/6013